formats/json/commonMain/src/kotlinx/serialization/json/internal/JsonReader.kt

```diff
@@ -120,202 +123,314 @@ internal class JsonReader(private val source: String) {
-            if (source[currentPosition] == STRING_ESC) {
-                appendRange(source, lastPosition, currentPosition)
-                val newPosition = appendEsc(source, currentPosition + 1)
-                currentPosition = newPosition
-                lastPosition = newPosition
+
+        return consumeKeyString()
+    }
+
+    private fun consumeString(startPosition: Int, current: Int): String {
+        var currentPosition = current
+        var lastPosition = startPosition
+        val source = source
+        var char = source[currentPosition] // Avoid two range checks visible in the profiler
+        while (char != STRING) {
+            if (char == STRING_ESC) {
+                currentPosition = appendEscape(lastPosition, currentPosition)
+                lastPosition = currentPosition
             } else if (++currentPosition >= source.length) {
                 fail("EOF", currentPosition)
             }
+            char = source[currentPosition]
         }
-        if (lastPosition == startPosition + 1) {
+
+        val string = if (lastPosition == startPosition) {
             // there was no escaped chars
-            offset = lastPosition
-            this.length = currentPosition - lastPosition
+            source.substring(lastPosition, currentPosition)
         } else {
             // some escaped chars were there
-            appendRange(source, lastPosition, currentPosition)
-            this.offset = -1
+            decodedString(lastPosition, currentPosition)
         }
         this.currentPosition = currentPosition + 1
-        tokenClass = TC_STRING
+        return string
+    }
+
+    private fun appendEscape(lastPosition: Int, current: Int): Int {
+        escapedString.append(source, lastPosition, current)
+        return appendEsc(current + 1)
+    }
+
+    private fun decodedString(lastPosition: Int, currentPosition: Int): String {
+        appendRange(lastPosition, currentPosition)
+        val result = escapedString.toString()
+        escapedString.setLength(0)
+        return result
     }
 
-    private fun appendEsc(source: String, startPosition: Int): Int {
+    private fun takePeeked(): String {
+        return peekedString!!.also { peekedString = null }
+    }
+
+    // Allows to consume unquoted string
+    fun consumeStringLenient(): String {
+        if (peekedString != null) {
+            return takePeeked()
+        }
+        var current = skipWhitespaces()
+        // Skip leading quotation mark
+        val token = charToTokenClass(source[current])
+        if (token == TC_STRING) {
+            return consumeString()
+        }
+
+        if (token != TC_OTHER) {
+            fail("Expected beginning of the string, but got ${source[current]}")
+        }
+        while (current < source.length && charToTokenClass(source[current]) == TC_OTHER) {
+            ++current
+        }
+        val result = source.substring(currentPosition, current)
```

Comment:
We indeed could have done it, but this is beyond the scope of this issue, I'd prefer separating performance improvements and actual semantic changes

Reply:
As far as I understand, current json parser supports escape sequences in lenient mode (because they're handled by the same function), so it may break things for some users. Can't you reuse `consumeString` instead of substring?

Reply:
No, it doesn't.

Previously, unquoted strings were handled via `nextToken` -> `nextLiteral`
Quoted strings in lenient mode were handled via `nextToken` -> `nextString`

It means that quoted strings with escapes worked properly in lenient mode, while unquoted were not.
You can see it by running
```
@Serializable
data class StringData(val s: String)
@Test
fun foo() {
    val json = """{"s": abc\nc}"""
    println(Json{ isLenient = true }.decodeFromString<StringData>(json))
}
```

in `dev` branch.

This behaviour is left intact in new parser